### ISSUE TITLE
discussion: migrate from DeepSpeed to FSDP2

### ISSUE THREAD

**Issue by gau-nernst** [allenai/open-instruct#1095]
FSDP2 is basically a PyTorch-native version of ZeRO-3: https://docs.pytorch.org/tutorials/intermediate/FSDP_tutorial.html
- FSDP2 has better composability with `torch.compile()` as guaranteed by the PyTorch team (also demonstrated in [torchtitan](https://github.com/pytorch/torchtitan) and [torchtune](https://github.com/meta-pytorch/torchtune). I'm not sure what's the status of DeepSpeed + torch.compile)
- In terms speed and memory, I don't have much prior experience with DeepSpeed so can't compare.
- Removing DeepSpeed dependency is also a benefit I suppose, if FSDP2 is not worse than DeepSpeed.
- However, switching to DeepSpeed means that there is no option to run the equivalence of ZeRO-1 and ZeRO-2.

I'm mainly looking at `grpo_fast.py` at the moment. From my brief understanding, here are the main code changes:
- `PolicyTrainerRayProcess.from_pretrained()` - model creation
- `PolicyTrainerRayProcess.broadcast_to_vllm()` - all-gather model weights to broadcast to vLLM
- `PolicyTrainerRayProcess.save_model()` - all-gather model weights for saving
- `PolicyTrainerRayProcess.train()` - training loop. Replace DeepSpeed API (i.e. `self.model.backward(loss)`, `model.step()`) with FSDP2 API (i.e. standard `loss.backward()`, `optim.step()`, `optim.zero_grad()`)

Opening an issue here mainly for discussion first.

cc @finbarrtimbers @hamishivi

**Comment by finbarrtimbers** [comment]
You're right to focus on `grpo_fast.py`. I am planning to refactor `grpo_fast.py` into a more general script that will let us do PPO and whatever other algorithms we want. 

The only question that I have is how FSDP2 works with multi-node. Do we lose any multi-node capabilities by switching to FSDP2? 

I know that FSDP2 is what the pretraining team uses (in [`OLMo-core`](https://github.com/allenai/OLMo-core)), so it should be fine.

**Comment by finbarrtimbers** [comment]
@gau-nernst to keep you updated, there's a good chance that we're going to switch to use a different trainer as part of https://github.com/allenai/open-instruct/issues/1103, in which case it might be worth holding off on this refactor.

The leading contenders are: [OLMo-core](https://github.com/allenai/OLMo-core) and [Megatron](https://github.com/NVIDIA/Megatron-LM). If you have thoughts, would appreciate them.

**Comment by hamishivi** [comment]
Closing since we decided on swapping to olmo-core for training! Please re-open or open a new issue if you have issues. Olmo-core uses FSDP and should make training much more efficient in general.
